### ISSUE TITLE
Mistake in logic for worker shutdown

### ISSUE THREAD

**Issue by pietdevaere** [mami-project/pathspider#92]
There is a mistake in the worker shutdown prodcedure.

# Intro
Once a worker has received the `SHUTDOWN SENTINEL`, it will go in to non active mode, and execute the following code in a loop: 

```
            else: # not worker_active, spin the semaphores
                self.sem_config_zero.acquire()
                time.sleep(QUEUE_SLEEP)
                with self.active_worker_lock:
                    if self.active_worker_count <= 0:
                        #self._worker_state[worker_number] = "shutdown_complete"
                        break
                self.sem_config_one_rdy.release()
                self.sem_config_one.acquire()
                # self._worker_state[worker_number] = "shutdown_1"
                time.sleep(QUEUE_SLEEP)
                self.sem_config_zero_rdy.release()
```

# Now, during the shutdown procedure:
When some workers arrive at the `if self.active_worker_count <= 0:` slghtly after the others, this condition might already be true, and they will shutdown.  
However, the other workers (who where there earlier) are now waiting for `self.sem_config_one.acquire()`, but because only a subset is the workers is still active, this means that `self.sem_config_one_rdy.release()` will not be called `worker_count` number of times.  
Thus the workers will wait forever.

# Reproducing
This bug was rather tricky to reproduce, as it only occurs sometimes. I have not been able to reproduce it in the `master` branch yet, but when running the `http_get` branch, with [this input file](https://devae.re/f/eth/pathspider/modern_times_1k.csv) it occurs about every one in two runs.

I think the reason that it shows up in `http_get` is because performing an HTTP GET request makes timing much less predictable.

# A note about certainty
Because it takes so long to reproduce this bug, it took me a *loooong* time to track it down. However, I am about 99% certain that the problem is as I described above.

**Comment by pietdevaere** [comment]
I wrote a fix, now running tests. This will have to run for a while as the bug is (or hopefully was) relatively hard to reproduce

**Comment by pietdevaere** [comment]
fixed in 836890cfb920136ec503078bb118c52cc754f56e
